### ISSUE TITLE
EnsureCreatedAsync() performance downgrade after upgrading Microsoft.Data.SqlClient from version 5.1.1 to 5.1.4(5.1.1-->5.1.4)

### ISSUE THREAD

**Issue by akhiltyagi98** [dotnet/SqlClient#2542]
### Describe the bug
I am using Microsoft.EntityFrameworkCore localDB SQLServer to run test cases, I use below commands in each test cases 

 await context.Database.EnsureDeletedAsync().ConfigureAwait(false);
 await context.Database.EnsureCreatedAsync().ConfigureAwait(false);

But after upgrading Microsoft.Data.SqlClient package version from 5.1.1 to 5.1.4(5.1.1-->5.1.4), EnsureCreatedAsync()  performance downgraded, can you please help me understand this ?

### To reproduce
Please use this code to check

```c#
using Microsoft.EntityFrameworkCore;
using Moq;
using Xunit;

namespace POCSQLClientPerformence.Tests
{
    public class SQLClientPerformenceDB
    {
        private IDbContextFactory<TestDBContext> factory;

        public SQLClientPerformenceDB()
        {
            CreateDbAsync().GetAwaiter().GetResult();
        }

        async Task CreateDbAsync()
        {
            Func<TestDBContext> CreateContext = () =>
            {
                var optionsBuilder = new DbContextOptionsBuilder<TestDBContext>()
                    .UseSqlServer(@$"Server=(localdb)\MSSQLLocalDb;Database=ABCTestDb;Trusted_Connection=True;MultipleActiveResultSets=true");

                return new TestDBContext(optionsBuilder.Options);
            };

            var factoryMock = new Mock<IDbContextFactory<TestDBContext>>();
            factoryMock.Setup(it => it.CreateDbContextAsync(It.IsAny<CancellationToken>())).ReturnsAsync(CreateContext);

            this.factory = factoryMock.Object;
            using var context = await this.factory.CreateDbContextAsync(CancellationToken.None).ConfigureAwait(false);

            await context.Database.EnsureDeletedAsync().ConfigureAwait(false);
            await context.Database.EnsureCreatedAsync().ConfigureAwait(false);
        }

        [Fact]
        public async Task FirstTest()
        {
            // Arrange
            using var context = await this.factory.CreateDbContextAsync(CancellationToken.None).ConfigureAwait(false);

            // Assert - Validate the results or behavior
            Assert.NotNull(context); // Just a dummy assertion for demonstration
        }

        [Fact]
        public async Task SecoundTest()
        {
            // Arrange
            using var context = await this.factory.CreateDbContextAsync(CancellationToken.None).ConfigureAwait(false);

            // Assert - Validate the results or behavior
            Assert.NotNull(context); // Just a dummy assertion for demonstration
        }

        [Fact]
        public async Task ThirdTest()
        {
            // Arrange
            using var context = await this.factory.CreateDbContextAsync(CancellationToken.None).ConfigureAwait(false);

            // Assert - Validate the results or behavior
            Assert.NotNull(context); // Just a dummy assertion for demonstration
        }
    }
 @}
```
Microsoft.Data.SQLClient: 5.2.0(Latest) test case execution time:
<img width="960" alt="image" src="https://github.com/dotnet/SqlClient/assets/171185381/5ef067bc-2f14-4cce-a1d1-4906303398d9">

Microsoft.Data.SQLClient version 5.1.1 or later package test case execution time:
<img width="960" alt="image" src="https://github.com/dotnet/SqlClient/assets/171185381/4cd4e8d8-e44d-447e-9da0-af9f329457cc">

**Comment by ErikEJ** [comment]
Fixed in 5.2.1

**Comment by arellegue** [comment]
Please try again after the 5.2.1 Hotfix is published and please let us know if you are still having an issue.

**Comment by akhiltyagi98** [comment]
I did not find Microsoft.Data.SqlClient version 5.2.1 and 5.1.6 please check published version. https://www.nuget.org/packages/Microsoft.Data.SqlClient/5.2.0?_src=template#versions-body-tab

Could you please explain why this is happening?

**Comment by arellegue** [comment]
That feature you requested to be investigated is in EntityFramework package.  Could you open an issue with EntityFramework instead, please? 

https://github.com/dotnet/efcore/issues

**Comment by akhiltyagi98** [comment]
Butts due to Microsoft.Data.SqlClient version only you can check upgrading Microsoft.Data.SqlClient version from 5.1.1 to 5.1.4. you can observe time difference

**Comment by arellegue** [comment]
We will investigate and get back to you when we have an update. When the Hotfix 5.2.1 is released later this week, could you try that and let us know if you are still having an issue.

**Comment by arellegue** [comment]
Could you provide a runnable repro in a zip file, please?

**Comment by akhiltyagi98** [comment]
@arellegue provided in MS Teams.

**Comment by ajcvickers** [comment]
Duplicate of #33399

**Comment by arellegue** [comment]
I have investigated this issue and found the root cause to be in the introduction of transient fault handling which added the delay.
We will discuss how to appropriately implement a fix. We will let you know when the issue is fixed.

**Comment by arellegue** [comment]
This is related to fix in #1983.

**Comment by JRahnama** [comment]
@akhiltyagi98 htofix v5.2.1 is released. Can you test with that version?

**Comment by akhiltyagi98** [comment]
@arellegue @JRahnama JRahnama

Thanks Team issue has been fixed in new release 5.2.1. Please mark this as close.

**Comment by akhiltyagi98** [comment]
Fixed in hotfix release 5.2.1.
